rust/src/lib/nm/profile_search.rs

```diff
@@ -0,0 +1,367 @@
+// SPDX-License-Identifier: Apache-2.0
+
+use std::collections::HashMap;
+use std::rc::Rc;
+
+use super::nm_dbus::{NmActiveConnection, NmConnection, NmIfaceType};
+use crate::{BaseInterface, InterfaceIdentifier, NetworkState};
+
+// Using &str for UUID or name will complex the code due to ownership,
+// Considering this duplication on UUID and name only created once at
+// `Self::new()` stage, it is acceptable cost.
+//
+// Using NmIfaceType instead of &NmIfaceType in order to matching veth
+// interface against ethernet NmConnection, we need to duplicate NmConnection
+// both to Veth and Ethernet key.
+//
+// Using String instead of &str for interface name indexing because VLAN
+// connection can have auto-generated `connection.interface-name` when it is
+// empty.
+#[derive(Debug, Default)]
+pub(crate) struct ProfileSearch {
+    nm_saved_cons: HashMap<String, Rc<NmConnection>>,
+    nm_applied_cons: HashMap<String, Rc<NmConnection>>,
+    // UUID to NmActiveConnection
+    nm_acs: HashMap<String, Rc<NmActiveConnection>>,
+    // (name, nm_iface_type) to Rc<NmActiveConnection>
+    // Veth will also stored into `NmIfaceType::Ethernet`
+    nm_acs_index: HashMap<(String, NmIfaceType), Rc<NmActiveConnection>>,
+
+    // (name, nm_iface_type) to Rc<NmConnection>
+    // Veth will also stored into `NmIfaceType::Ethernet`
+    nm_applied_conns_index: HashMap<(String, NmIfaceType), Rc<NmConnection>>,
+
+    // (name, nm_iface_type) to Rc<NmConnection>
+    // Veth will also stored into `NmIfaceType::Ethernet`
+    nm_saved_conns_index: HashMap<(String, NmIfaceType), Vec<Rc<NmConnection>>>,
+
+    // (mac_upper_case, nm_iface_type) to Vec<Rc<NmConnection>> for mac
+    // identifier Veth will also stored into `NmIfaceType::Ethernet`
+    nm_saved_conns_mac: HashMap<(String, NmIfaceType), Vec<Rc<NmConnection>>>,
+
+    cur_state: NetworkState,
+}
+
+impl ProfileSearch {
```

Comment:
Suggestion: rename to NmConnectionMatcher.

Reply:
Changed.